src/Passes/Source/Rules/Factory.cpp

```diff
@@ -0,0 +1,334 @@
+      , result_alloc_manager_{AllocationManager::createNew()}
+    {
+    }
+
+    RuleFactory::AllocationManagerPtr RuleFactory::qubitAllocationManager() const
+    {
+        return qubit_alloc_manager_;
+    }
+
+    RuleFactory::AllocationManagerPtr RuleFactory::resultAllocationManager() const
+    {
+        return result_alloc_manager_;
+    }
+
+    void RuleFactory::removeFunctionCall(String const& name)
+    {
+        addRule({callByNameOnly(name), deleteInstruction()});
+    }
+
+    void RuleFactory::useStaticQubitArrayAllocation()
+    {
+        // TODO(tfr): Consider using weak pointers
+        auto qubit_alloc_manager = qubit_alloc_manager_;
+
+        /// Allocation
+        auto allocation_replacer =
+            [qubit_alloc_manager](Builder&, Value* val, Captures& cap, Replacements& replacements) {
+                auto cst = llvm::dyn_cast<llvm::ConstantInt>(cap["size"]);
+                if (cst == nullptr)
+                {
+                    return false;
+                }
+
+                auto llvm_size = cst->getValue();
+                auto name      = val->getName().str();
+                qubit_alloc_manager->allocate(name, llvm_size.getZExtValue());
+
+                replacements.push_back({llvm::dyn_cast<Instruction>(val), nullptr});
+                return true;
+            };
+
+        addRule({call("__quantum__rt__qubit_allocate_array", "size"_cap = _), allocation_replacer});
+
+        /// Array access replacement
+        auto access_replacer =
+            [qubit_alloc_manager](Builder& builder, Value* val, Captures& cap, Replacements& replacements) {
+                // Getting the type pointer
+
+                auto ptr_type = llvm::dyn_cast<llvm::PointerType>(val->getType());
+                if (ptr_type == nullptr)
+                {
+                    return false;
+                }
+
+                // Get the index and testing that it is a constant int
+                auto cst = llvm::dyn_cast<llvm::ConstantInt>(cap["index"]);
+                if (cst == nullptr)
+                {
+                    // ... if not, we cannot perform the mapping.
+                    return false;
+                }
+
+                // Computing the index by getting the current index value and offseting by
+                // the offset at which the qubit array is allocated.
+                auto llvm_size = cst->getValue();
+                auto offset    = qubit_alloc_manager->getOffset(cap["arrayName"]->getName().str());
+
+                // Creating a new index APInt that is shifted by the offset of the allocation
+                auto idx = llvm::APInt(llvm_size.getBitWidth(), llvm_size.getZExtValue() + offset);
+
+                // Computing offset
+                auto new_index = llvm::ConstantInt::get(builder.getContext(), idx);
+
+                // TODO(tfr): Understand what the significance of the addressspace is in relation to the
```

Comment:
I am not sure this todo still applies, if the static qubit allocations are using the default address space as well.

Reply:
No it does not apply any longer - we covered that.

Reply:
It is not!